### PR TITLE
add batched changes to trie

### PR DESCRIPTION
**Pull request by BeniaminDrasovean**
## Reasoning behind the pull request
- All the trie changes are added to a batch before being added to the trie. When the batch is commited to the trie, the changes are added one by one. 
  
## Proposed changes
- In order to reduce the time spent on saving the changes from the batch to the trie, add all the batched data to the trie in one go. 
- The next step will be to add the data to the trie using goroutines

## Testing procedure
- Normal testing procedure

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: trie/patriciaMerkleTrie.go
Comment (ssd04) [inline trie/patriciaMerkleTrie.go:1]: unit tests for the changes here?

File: trie/branchNode.go [hunk 1]
```diff
@@ -443,61 +443,139 @@ func (bn *branchNode) getNext(key []byte, db common.TrieStorageInteractor) (node
...
+		return bn, modifiedHashes, nil
 	}
 
-	if bn.children[childPos] == nil {
-		return bn.insertOnNilChild(newData, childPos)
+	return nil, emptyHashes, nil
+}
+
+// the prerequisite for this to work is that the data is already sorted
+func splitDataForChildren(newData []core.TrieData) ([][]core.TrieData, error) {
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/branchNode.go:498]: Done

File: trie/extensionNode.go [hunk 2]
```diff
@@ -354,16 +354,50 @@ func (en *extensionNode) insert(newData core.TrieData, db common.TrieStorageInte
...
 
 	// Otherwise branch out at the index where they differ.
-	return en.insertInNewBn(newData, keyMatchLen)
+	return en.insertInNewBn(newData, db, keyMatchLen, index)
 }
 
-func (en *extensionNode) insertInSameEn(newData core.TrieData, keyMatchLen int, db common.TrieStorageInteractor) (node, [][]byte, error) {
-	newData.Key = newData.Key[keyMatchLen:]
+func getMinKeyMatchLen(newData []core.TrieData, enKey []byte) (int, int) {
+	keyMatchLen := len(enKey)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/extensionNode.go:361]: Done.

File: trie/branchNode_test.go [hunk 3]
```diff
@@ -1511,3 +1530,464 @@ func TestBranchNode_revertChildrenVersionSliceIfNeeded(t *testing.T) {
...
+		{
+			Key:   []byte{6, 10, 11, 16},
+			Value: []byte("value3"),
+		},
+		{
+			Key:   []byte{16},
+			Value: []byte("value4"),
+		},
+	}
+	err := bn.commitDirty(0, 5, db, db)
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/branchNode_test.go:1868]: Will add in a future PR, when I will refactor all trie tests

File: trie/branchNode_test.go [hunk 4]
```diff
@@ -1511,3 +1530,464 @@ func TestBranchNode_revertChildrenVersionSliceIfNeeded(t *testing.T) {
...
+	assert.Nil(t, err)
+	assert.False(t, bn.dirty)
+
+	newNode, modifiedHashes, err := bn.insert(newData, db)
+	assert.Nil(t, err)
+	assert.Equal(t, 2, len(modifiedHashes))
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/branchNode_test.go:1874]: Done.

File: trie/patriciaMerkleTrie.go
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:1]: Will add in a future PR.

File: trie/patriciaMerkleTrie.go [hunk 5]
```diff
@@ -172,56 +172,73 @@ func (tr *patriciaMerkleTrie) updateTrie() error {
...
+	if tr.root == nil {
+		newRoot, err := newLeafNode(sortedDataForInsertion[0], tr.marshalizer, tr.hasher)
 		if err != nil {
 			return err
 		}
 
-		if check.IfNil(newRoot) {
-			continue
+		tr.root = newRoot
+		sortedDataForInsertion = sortedDataForInsertion[1:]
```
[the remark this answers is not in the document]
Comment (BeniaminDrasovean) [reply trie/patriciaMerkleTrie.go:195]: Yes. On L189, `sortedDataForInsertion[0]` is added to the trie. If the `sortedDataForInsertion` had only one val, then this func will return on L198. Added a test to prove it works.

**Review by sasurobert — APPROVED** [review]

**Review by ssd04 — APPROVED** [review]

**Pull request merged**